### DISCUSSION TITLE
How to deduce the private and public key-lengths from the parsed ASN.1 structure?

### DISCUSSION THREAD

**Discussion by Hussain1811** [open-quantum-safe/oqs-provider#415] [Q&A]
I've generated a dilithium key as `openssl genpkey -algorithm Dilithium2 -provider oqsprovider > dilithium2_priv.pem`
Later, parsed the ASN.1 structure using `openssl asn1parse -in dilithium2_priv.pem`
I could see that the octet string containing the actual key is of length 3844.

But, dilithium2 private key-size is `2528 `according to `OQS_SIG_dilithium_2_length_secret_key`.

Am i missing anything?

**Comment by Hussain1811** [comment]
I am able to interpret the public key-length from the ASN.1 structure.
But, unable to deduce for a private key.

Can anybody here guide on the same?

**Comment by baentsch** [comment]
> I could see that the octet string containing the actual key is of length 3844.

> But, dilithium2 private key-size is 2528 according to OQS_SIG_dilithium_2_length_secret_key.

> Am i missing anything?

Yes: `oqsprovider` by [default encodes the public key as part of the private key](https://github.com/open-quantum-safe/oqs-provider/blob/main/CONFIGURE.md#nopubkey_in_privkey) as `liboqs` [does not contain logic to re-create a public from a private key](https://github.com/open-quantum-safe/liboqs/issues/1802).

**Comment by baentsch** [comment]
Ah, only now saw you already closed this, @Hussain1811 . Then you probably already found the answer. For posterity it'd be nice if you'd post it here so others may benefit from it (or completely delete it if you don't deem this relevant for others).
